### PR TITLE
Separate the execution logic

### PR DESCRIPTION
**Pull request by esunar**
Extract the execution logic to a simple interface
Created a simple ExecutorFactory
Created a simple SerialExecutor by using the cli.apply_plan 
Used SerialExecutor in cli

### CODE REVIEW

**Review by rgildein — COMMENTED** [review]
Why we are creating the BaseExecutor, ExecutorFactory and SerialExecutor? Why creating class object if it's not necessary? We create just more friction, and even pylint is telling us not to do so why even use pylint if we doing disable.

**Comment by esunar** [comment]
> Why we are creating the BaseExecutor, ExecutorFactory and SerialExecutor? Why creating class object if it's not necessary? We create just more friction, and even pylint is telling us not to do so why even use pylint if we doing disable.

At the time being, it seems over engineering. However it is a guideline to show how will the application evolve. For example in near future we will be using parallel execution. When we move to data plane we will be using both. When it get complicated it will be really hard to maintain in a module with bunch of functions.

File: cou/steps/execute.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+async def _apply_plan(plan: UpgradeStep, interactive: bool) -> None:
+    """Apply the plan for upgrade.
+
+    :param plan: Plan to be executed on steps.
+    :type plan: UpgradeStep
+    :param interactive:
+    :type interactive: bool
+    """
+    result = "X"
```
Comment (aieri) [inline cou/steps/execute.py:64]: nit: `X` looks at a first glance as a possible option. We could use an empty string here to represent 'no result'

File: setup.cfg [hunk 2]
```diff
@@ -6,19 +6,19 @@ description = Application for upgrading Charmed Openstack
 long_description = file: README.md
 long_description_content_type = text/markdown; charset=UTF-8; variant=GFM
 author = Canonical DevOps Centres
-license = GPLv3
+license = Apache-2.0
```
Comment (aieri) [inline setup.cfg:9]: A change in licence seems out of scope for this PR. Also: why?

Comment (aieri) [reply setup.cfg:9]: the headers of other files in this PR are also still mentioning the GPL

**Review by aieri — CHANGES_REQUESTED** [review]

File: setup.cfg (hunk 2 above)
Comment (esunar) [reply setup.cfg:9]: I'll remove that part and open another pr for license.

File: cou/cli.py [hunk 3]
```diff
@@ -188,7 +136,7 @@ async def entrypoint() -> None:
         print(analysis_result)
         upgrade_plan = await generate_plan(analysis_result)
         if args.run:
-            await apply_plan(upgrade_plan, args.non_interactive)
+            await execute(upgrade_plan, not args.non_interactive)
```
[the remark this answers is not in the document]
Comment (esunar) [reply cou/cli.py]: used the second approach

File: cou/steps/execute.py (hunk 1 above)
Comment (esunar) [reply cou/steps/execute.py]: removed `X`

File: setup.cfg (hunk 2 above)
Comment (esunar) [reply setup.cfg]: removed

File: cou/steps/execute.py [hunk 4]
```diff
@@ -61,8 +60,8 @@ async def _apply_plan(plan: UpgradeStep, interactive: bool) -> None:
     :param interactive:
     :type interactive: bool
     """
-    result = "X"
-    while result.casefold() not in AVAILABLE_OPTIONS:
+    result = ""
+    while result == "" or result.casefold() not in AVAILABLE_OPTIONS:
```
Comment (aieri) [inline cou/steps/execute.py:64]: ah yes, because an empty string is considered to always be part of a string. You could (non blocking) define `AVAILABLE_OPTIONS` as a list and get rid of the `result == ""` comparison.

**Review by aieri — DISMISSED** [review]
LGTM w/one optional suggestion

File: cou/steps/execute.py (hunk 4 above)
Comment (esunar) [reply cou/steps/execute.py]: done
